sszgen/tag_parser.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"text/scanner"
+)
+
+type tokenState int
+
+const (
+	tsBegin tokenState = iota
+	tsLabel
+	tsValue
+	tsCloseTick
+)
+
+type TagParser struct {
+	sc scanner.Scanner
+	buffer string
+}
+
+func (tp *TagParser) Init(tag string) {
+	sr := strings.NewReader(tag)
+	tp.sc = scanner.Scanner{}
+	tp.sc.Init(sr)
+	tp.sc.Filename = "tag"
+	tp.sc.Mode ^= scanner.ScanRawStrings
+}
+
+func (tp TagParser) GetSSZTags() map[string]string {
+	var labelStr string
+	var state tokenState
+	tags := make(map[string]string)
+	for tok := tp.sc.Scan(); tok != scanner.EOF; tok = tp.sc.Scan() {
+		if state == tsCloseTick {
+			panic("undefined behavior when scanning beyond the end of the tag")
+		}
+		txt := tp.sc.TokenText()
+		switch txt {
+		case "`":
+			if state == tsLabel {
+				state = tsCloseTick
+				continue
+			}
+			if state == tsBegin {
+				state = tsLabel
+				continue
+			}
+		case ":":
+			if state == tsLabel {
+				state = tsValue
+				continue
+			}
+		case "\"":
+			continue
+		default:
+			if state == tsValue {
+				tags[labelStr] = trimTagQuotes(string(txt))
+				state = tsLabel
+				labelStr = ""
+				continue
+			}
+			if state == tsLabel {
+				labelStr += string(txt)
+				continue
+			}
+		}
+	}
+	return tags
+}
+
+// cannot compare untyped nil to typed nil
+// this value gives us a nil with type of *int
+// to compare to ssz-size = '?' values
+var nilInt *int
+
+func extractSSZDimensions(tag string) ([]*SSZDimension, error) {
+	tp := &TagParser{}
+	tp.Init(tag)
+	// parse the ssz-max and ssz-size key/value pairs out of the tag
+	tags := tp.GetSSZTags()
+	sszSizes, sizeDefined := tags["ssz-size"]
+	sszMax, maxDefined:= tags["ssz-max"]
+	if !sizeDefined && !maxDefined {
+		return nil, fmt.Errorf("No ssz-size or ssz-max tags found for element. tag=%s", tag)
+	}
+
+	// split each tag by ",". each position in the csv represents a dimension of an n-dimensional array
+	sizeSplit := strings.Split(sszSizes, ",")
+	maxSplit := strings.Split(sszMax, ",")
+	// find the largest of the two dimensions. for backward compat we'll be permissive and let them be uneven
+	ndims := len(sizeSplit)
+	if len(maxSplit) > len(sizeSplit) {
+		ndims = len(maxSplit)
+	}
+	dims := make([]*SSZDimension, ndims)
+	for i := 0; i < ndims; i++ {
+		var szi, mxi string
+		if len(sizeSplit) > i {
+			szi = sizeSplit[i]
+		}
+		if len(maxSplit) > i {
+			mxi = maxSplit[i]
+		}
+		if szi == "?" && mxi == "?" {
+			return nil, fmt.Errorf("At dimension %d both ssz-size and ssz-max had a '?' value, tag=%s", i, tag)
```

Comment:
Maybe a more informational error to someone using the generator could be helpful. Something that explains that if you're trying to specify a multi-dimensional list, you should do x...otherwise, do y

Reply:
ptal